File: app/src/main/java/com/hiddenramblings/tagmo/eightbit/net/JSONExecutor.java
/*
 * ====================================================================
 * Copyright (c) 2012-2022 AbandonedCart.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * For the purpose of this license, the phrase "TagMo labels" shall
 * be used to refer to the labels "8-bit Dream", "TwistedUmbrella",
 * "TagMo" and "AbandonedCart" and these labels should be considered
 * the equivalent of any usage of the aforementioned phrase.
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All materials mentioning features or use of this software and
 *    redistributions of any form whatsoever must display the following
 *    acknowledgment unless made available by tagged, public "commits":
 *    "This product includes software developed for TagMo by AbandonedCart"
 *
 * 4. The TagMo labels must not be used in any form to endorse or promote
 *    products derived from this software without prior written permission.
 *    For written permission, please contact enderinexiledc@gmail.com
 *
 * 5. Products derived from this software may not be called by the TagMo labels
 *    nor may these labels appear in their names or product information without
 *    prior written permission of AbandonedCart.
 *
 * THIS SOFTWARE IS PROVIDED BY AbandonedCart AND TagMo ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * The license and distribution terms for any publicly available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution license
 * [including the GNU Public License.] Content not subject to these terms is
 * subject to to the terms and conditions of the Apache License, Version 2.0.
 */

package com.hiddenramblings.tagmo.eightbit.net;

import android.app.Activity;

import com.hiddenramblings.tagmo.GlideTagModule;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.AmiiboManager;
import com.hiddenramblings.tagmo.browser.BrowserActivity;
import com.hiddenramblings.tagmo.browser.Preferences;
import com.hiddenramblings.tagmo.charset.CharsetCompat;
import com.hiddenramblings.tagmo.eightbit.io.Debug;
import com.hiddenramblings.tagmo.security.SecurityHandler;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.concurrent.Executors;

import javax.net.ssl.HttpsURLConnection;

public class JSONExecutor {

    ResultListener listener;

    public JSONExecutor(Activity activity, String server, String path) {
        new SecurityHandler(activity, new SecurityHandler.ProviderInstallListener() {
            @Override
            public void onProviderInstalled() {
                RetrieveJSON(server, path);
            }

            @Override
            public void onProviderInstallException() {
                RetrieveJSON(server, path);
            }

            @Override
            public void onProviderInstallFailed() {
                if (activity instanceof BrowserActivity) {
                    new Preferences(activity.getApplicationContext())
                            .image_network(GlideTagModule.IMAGE_NETWORK_NEVER);
                    activity.runOnUiThread(() ->
                            ((BrowserActivity) activity).getSettings().notifyChanges()
                    );
                }
                listener.onResults(null);
            }
        });
    }

    private HttpsURLConnection fixServerLocation(URL url) throws IOException {
        HttpsURLConnection urlConnection = (HttpsURLConnection) url.openConnection();
        urlConnection.setRequestMethod("GET");
        urlConnection.setUseCaches(false);
        urlConnection.setDefaultUseCaches(false);
        return urlConnection;
    }

    public void RetrieveJSON(String server, String path) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                HttpsURLConnection conn = (HttpsURLConnection)
                        new URL(server + path).openConnection();
                conn.setRequestMethod("GET");
                conn.setUseCaches(false);
                conn.setDefaultUseCaches(false);

                int statusCode = conn.getResponseCode();
                if (statusCode == HttpsURLConnection.HTTP_MOVED_PERM) {
                    String address = conn.getHeaderField("Location");
                    conn.disconnect();
                    conn = fixServerLocation(new URL(address));
                    statusCode = conn.getResponseCode();
                } else if (statusCode != HttpsURLConnection.HTTP_OK && isRenderAPI(conn)) {
                    conn.disconnect();
                    conn = fixServerLocation(new URL(AmiiboManager.AMIIBO_API + "amiibo/"));
                    statusCode = conn.getResponseCode();
                }

                if (statusCode != HttpsURLConnection.HTTP_OK) {
                    conn.disconnect();
                    return;
                }

                InputStream in = conn.getInputStream();
                BufferedReader streamReader = new BufferedReader(
                        new InputStreamReader(in, CharsetCompat.UTF_8));
                StringBuilder responseStrBuilder = new StringBuilder();

                String inputStr;
                while (null != (inputStr = streamReader.readLine()))
                    responseStrBuilder.append(inputStr);

                listener.onResults(responseStrBuilder.toString());
                streamReader.close();
                in.close();
                conn.disconnect();
            } catch (IOException e) {
                Debug.Warn(e);
            }
        });
    }

    private boolean isRenderAPI(HttpsURLConnection urlConnection) {
        String render = AmiiboManager.RENDER_RAW + "database/amiibo.json";
        return render.equals(urlConnection.getURL().toString());
    }

    public interface ResultListener {
        void onResults(String result);
    }

    public void setResultListener(ResultListener listener) {
        this.listener = listener;
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/eightbit/net/JSONExecutor.kt
/*
 * ====================================================================
 * Copyright (c) 2012-2023 AbandonedCart
 *
 * LICENSE.AbandonedCart
 * ====================================================================
 */
package com.hiddenramblings.tagmo.eightbit.net

import android.app.Activity
import com.hiddenramblings.tagmo.BrowserActivity
import com.hiddenramblings.tagmo.GlideTagModule
import com.hiddenramblings.tagmo.Preferences
import com.hiddenramblings.tagmo.amiibo.AmiiboManager
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.security.SecurityHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.nio.charset.StandardCharsets
import javax.net.ssl.HttpsURLConnection

class JSONExecutor(activity: Activity, server: String, path: String? = null) {

    var jsonListener: ResultListener? = null
    var dbListener: DatabaseListener? = null
    private var hasPendingResult = false
    private var pendingResult: String? = null
    private var pendingException: Exception? = null

    init {
        SecurityHandler(activity, object : SecurityHandler.ProviderInstallListener {
            override fun onProviderInstalled() {
                retrieveJSON(server, path)
            }

            override fun onProviderInstallException() {
                retrieveJSON(server, path)
            }

            override fun onProviderInstallFailed() {
                Preferences(activity.applicationContext).imageNetwork(
                    GlideTagModule.IMAGE_NETWORK_NEVER
                )
                if (activity is BrowserActivity)
                    CoroutineScope(Dispatchers.Main).launch { activity.settings?.notifyChanges() }
                onResults(null)
            }
        })
    }

    @get:Throws(IOException::class)
    private val URL.asConnection get() : HttpsURLConnection {
        return (openConnection() as HttpsURLConnection).apply {
            requestMethod = "GET"
            useCaches = false
            defaultUseCaches = false
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
        }
    }

    fun retrieveJSON(server: String, path: String?) {
        CoroutineScope(Dispatchers.IO).launch(Dispatchers.IO) {
            val url = path?.let { "$server/$path" } ?: server
            var lastException: IOException? = null
            repeat(MAX_ATTEMPTS) { attempt ->
                try {
                    onResults(readJSON(url))
                    return@launch
                } catch (e: HttpStatusException) {
                    Debug.warn(e)
                    onException(e)
                    return@launch
                } catch (e: IOException) {
                    lastException = e
                    Debug.warn(e)
                    if (attempt < MAX_ATTEMPTS - 1) {
                        delay(RETRY_DELAY_MS * (attempt + 1))
                    }
                }
            }
            lastException?.let(::onException)
        }
    }

    private fun readJSON(url: String): String {
        var conn: HttpsURLConnection = URL(url).asConnection
        try {
            var statusCode = conn.responseCode
            if (statusCode == HttpsURLConnection.HTTP_MOVED_PERM
                || statusCode == HttpsURLConnection.HTTP_MOVED_TEMP
            ) {
                val address = conn.getHeaderField("Location")
                    ?: throw IOException("Redirect without a Location header")
                conn.disconnect()
                conn = URL(address).asConnection
                statusCode = conn.responseCode
            } else if (statusCode != HttpsURLConnection.HTTP_OK && isRawJSON(conn)) {
                conn.disconnect()
                conn = URL("${AmiiboManager.AMIIBO_RAW}/database/amiibo.json").asConnection
                statusCode = conn.responseCode
            }
            if (statusCode != HttpsURLConnection.HTTP_OK) {
                throw HttpStatusException(statusCode)
            }
            return BufferedReader(
                InputStreamReader(conn.inputStream, StandardCharsets.UTF_8)
            ).use { it.readText() }
        } finally {
            conn.disconnect()
        }
    }

    private class HttpStatusException(statusCode: Int) : IOException(
        "Unexpected HTTP response: $statusCode"
    )

    private fun onResults(result: String?) {
        dbListener?.onResults(result) ?: jsonListener?.onResults(result) ?: run {
            pendingResult = result
            hasPendingResult = true
        }
    }

    private fun onException(e: Exception) {
        dbListener?.onException(e) ?: jsonListener?.onException(e) ?: run {
            pendingException = e
        }
    }

    private fun dispatchPending() {
        if (hasPendingResult) {
            hasPendingResult = false
            onResults(pendingResult)
            pendingResult = null
        }
        pendingException?.let {
            pendingException = null
            onException(it)
        }
    }

    private fun isRawJSON(url: String): Boolean {
        return url.startsWith(AmiiboManager.RENDER_RAW)
    }

    private fun isRawJSON(urlConnection: HttpsURLConnection): Boolean {
        return isRawJSON(urlConnection.url.toString())
    }

    interface ResultListener {
        fun onResults(result: String?)
        fun onException(e: Exception)
    }

    fun setResultListener(listener: ResultListener?) {
        jsonListener = listener
        dispatchPending()
    }

    interface DatabaseListener {
        fun onResults(result: String?)
        fun onException(e: Exception)
    }

    fun setDatabaseListener(listener: DatabaseListener?) {
        dbListener = listener
        dispatchPending()
    }

    companion object {
        private const val CONNECT_TIMEOUT_MS = 15_000
        private const val READ_TIMEOUT_MS = 30_000
        private const val MAX_ATTEMPTS = 3
        private const val RETRY_DELAY_MS = 1_000L
    }
}
